crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -312,5 +392,117 @@ fn check_limits(limits: &Limits, nodes: u32, depth: u32, pos: Option<Pos>) -> Se
+            ),
+            pos,
+        ));
+    }
+
     Ok(())
 }
+
+/// Given a node, estimate the number of output nodes it will produce.
+fn estimate_output_nodes_for_curr_node(
+    f: &Positioned<Field>,
+    variables: &Variables,
+    parent_is_connection: bool,
+) -> Result<(u64, bool), ParseQueryError> {
+    let mut current_count = 1;
+    let mut is_connection = false;
+
+    // If the parent connection has 'first' or 'last' set, parent_is_connection is true.
+    // In that scenario, 'edges' and 'nodes' are normal, non-connection fields.
+    // Otherwise, this node is considered a connection and has count set to default_page_size.
+    if f.node.name.node == "edges" || f.node.name.node == "nodes" {
+        if !parent_is_connection {
+            current_count = DEFAULT_PAGE_SIZE;
+            is_connection = true;
+        }
+    } else {
+        // Check if the current node is a connection by checking if it has 'first' or 'last' selected as a field
+        let first_arg = f.node.get_argument("first");
+        let last_arg = f.node.get_argument("last");
+
+        match (first_arg, last_arg) {
+            (None, None) => { /* no action needed, use defaults */ }
+            // A value was provided for 'first' or 'last' or both
+            _ => {
+                let (first_count, last_count) =
+                    get_end_values_from_fields(first_arg, last_arg, variables)?;
+
+                // This follows the cursor spec for when both args are provided
+                // the resulting slice is the min of the two
+                let min_count = match (first_count, last_count) {
+                    (Some(first), Some(last)) => Some(std::cmp::min(first, last)),
+                    _ => first_count.or(last_count),
+                };
+
+                match min_count {
+                    Some(count) => {
+                        current_count = count;
+                        is_connection = true;
+                    }
+                    None => {
+                        is_connection = false;
+                    }
+                };
+            }
+        }
+    }
+
+    Ok((current_count, is_connection))
+}
+
+/// Given a node, check its args for 'first' and 'last', and return the values as
+/// a tuple of u64 notating how many to select from either end.
+fn get_end_values_from_fields(
+    first: Option<&Positioned<GqlValue>>,
+    last: Option<&Positioned<GqlValue>>,
+    variables: &Variables,
+) -> Result<(Option<u64>, Option<u64>), ParseQueryError> {
+    let first_count = first.map(|arg| extract_value(arg, variables)).transpose()?;
+    let last_count = last.map(|arg| extract_value(arg, variables)).transpose()?;
+    Ok((first_count, last_count))
+}
+
+/// Extracts the numerical value from a GraphQL value.
+/// The input can be a variable or a number.
+/// Its corresponding value must be a valid 64.
+/// If the value is a variable, it will look up the value in the variables map.
+fn extract_value(
```

Comment:
I was hoping this function would be prettier :P Seems rather complicated for getting those values.

Reply:
async-graphql exposes a ConstValue which is what we deal with in graphql. But when a request is given to the server its values come in the form of what is dubbed GqlValue here. The main difference is that GqlValue has an additional variant for Variable. Since it's possible for callers to provide a variable for first or last, we need to handle that scenario. And from there, we need to make sure that it is a Value::Number.

I suppose to make the code easier to read I can cast GqlValue -> ConstValue? it would just be an additional conversion instead of the lookup

Reply:
> First stab in the dark from me: Is there a slightly later phase we can hook into where variables have already been resolved etc, to avoid the complication from that?

I think parse_query is the best place to do these custom validations. The other hooks don't provide the necessary information, or are already in the process of resolving fields

Reply:
also, today i learned its totally fine to pass variables without a value defined if the field using the value is nullable